### PR TITLE
[LLVMGPU] Add subgroup size to target info and propagate it

### PR DESCRIPTION
**Pull request by kuhar**
This add a new parameter `subgroup_size =` to `hal.executable.export`, alongside the existing workgroup size. We do something similar on the SPIR-V codegen side.

Configure ROCm to use wave64 for GCN/CDNA and wave32 or wave64 for RDNA.
This is expected to improve performance on CDNA where we used to tune for wave32, which is not the native size for that hardware.

Check that the subgroup attribute is set in a few codegen pipelines across ROCm and CUDA.

-----

Co-authored-by: Quinn Dawkins

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 1]
```diff
@@ -805,11 +836,13 @@ static LogicalResult setWarpReductionConfig(func::FuncOp entryPoint,
   if (numDynamicReductionDims) {
     SmallVector<int64_t> reductionTileSizes(op.getNumLoops(), 0);
     // TODO: Don't hard code this.
-    reductionTileSizes[reductionDims[0]] = kCudaWarpSize;
+    reductionTileSizes[reductionDims[0]] =
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: Nit: I'd typically create a local variable for long expressions like `targetInfo.supportedSubgroupSizes` given it's used multiple times. Makes the code a bit nicer to read.

**Review by antiagainst — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 2]
```diff
@@ -160,6 +164,7 @@ static TargetInfo getCudaTargetInfo(func::FuncOp entryPoint) {
   TargetInfo info;
   // All the cuda target are assumed to have warp support.
   info.hasWarpShuffle = true;
+  info.supportedSubgroupSizes = {32};
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:167]: nit: Add a comment that the first subgroup size is "preferred" or "default" for most pipelines. Or, should we require that the subgroup sizes are always given in ascending order?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 3]
```diff
@@ -805,11 +836,12 @@ static LogicalResult setWarpReductionConfig(func::FuncOp entryPoint,
   if (numDynamicReductionDims) {
     SmallVector<int64_t> reductionTileSizes(op.getNumLoops(), 0);
     // TODO: Don't hard code this.
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: nit: You can remove this comment, it was referential to the cuda warp size.

**Review by qedawkins — APPROVED** [review]
Minor request, can you remove the reference to my github ID in the commit message before merging? If you want to can do co-authored or just remove it, but when people cherry-pick or rebase changes in local forks I start getting a bunch of pings for this mention.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp (hunk 2 above)
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:167]: Good point, I thought I had that comment already but I can't find it anymore. These are supposed to be listed in the order of preference, not necessarily monotonically.  For example, I don't think there's consensus on whether wave32 is better than wave64 on gfx11, so we may later swap the other of preference around.

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:167]: ah yes, the comment is there above

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:167]: Makes sense, preference sounds good to me.

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:167]: ah sorry, I missed it

**Pull request merged**